### DISCUSSION TITLE
Run_OpenLoop Error

### DISCUSSION THREAD

**Discussion by TheAfzalAmanullah** [borgestassio/Wind-Turbine-Control#4] [General]
Dear Sir,

I am getting an error after running the function Run_OpenLoop.m. The error are-

  FAST completed.

Error using Run_OpenLoop (line 30)
Error reported by S-function 'FAST_SFunc' in 'OpenLoop/FAST Nonlinear Wind Turbine/S-Function':
FAST_InitializeAll:SrvD_Init:ValidatePrimaryData:Yaw angle and rate are not commanded from Simulink model.
ValidatePrimaryData:Pitch angles are not commanded from Simulink model.
ValidatePrimaryData:Generator torque and power are not commanded from Simulink model.
ValidatePrimaryData:HSS brake is not commanded from Simulink model.
SrvD_Init:BladedInterface_Init:The dynamic library ..\..\CertTest\5MW_Baseline/ServoData/DISCON_win32.dll could not be loaded. Check that the file exists in the specified location and that it is
compiled for 64-bit applications

**Comment by borgestassio** [comment]
Just to clarify where this is coming from, which .fst files are you trying to run?
The instructions on this repo are focused on the 5MW model, if you changed the line #66 of another turbine (let's say the AWT27) it will not work.

**Reply by TheAfzalAmanullah** [reply]
I changed only in **NRELOffshrBsline5MW_Onshore_ServoDyn.dat** file and then run the **Run_OpenLoop.m** script file. Again I get the same error. 

Error using Run_OpenLoop (line 30)
Error reported by S-function 'FAST_SFunc' in 'OpenLoop/FAST Nonlinear Wind Turbine/S-Function':
FAST_InitializeAll:SrvD_Init:ValidatePrimaryData:Yaw angle and rate are not commanded from Simulink model.
ValidatePrimaryData:Pitch angles are not commanded from Simulink model.
ValidatePrimaryData:Generator torque and power are not commanded from Simulink model.
ValidatePrimaryData:HSS brake is not commanded from Simulink model.
SrvD_Init:BladedInterface_Init:The dynamic library ..\..\CertTest\5MW_Baseline/ServoData/DISCON_win32.dll could not be loaded. Check that the file exists in the specified location and that it is
compiled for 64-bit applications.

**Reply by borgestassio** [reply]
This error seems to be related to the 32bit DLL, which is odd since you mentioned that you're using a 64bit OS.
Could you try to run just the Test18.fst and share the output here?
You can edit the 'Run_OpenLoop.m' to run just Test18, to do so, you change the following line:

`for iTest = [18:18]`

This will make it easier to identify the problem.
Another option is to run the Test18.fst directly from the command line using the Fast V8 x64 executable.

**Reply by TheAfzalAmanullah** [reply]
I run the Test18.fst directly from the FAST_x64 executable. It run perfectly with no error. 
C:\Users\AFZAL\Desktop\FAST1\CertTest>FAST_x64 Test18.fst

**************************************************************************************************
FAST (v8.16.00a-bjj, 27-Jul-2016)

Copyright (C) 2016 National Renewable Energy Laboratory

This program comes with ABSOLUTELY NO WARRANTY. See the "license.txt" file distributed with this
software for details.
**************************************************************************************************

 Running FAST (v8.16.00a-bjj, 27-Jul-2016), compiled as a 64-bit application using single
 precision
 linked with NWTC Subroutine Library (v2.09.00, 23-Jul-2016)

 Heading of the FAST input file:
   FAST Certification Test #18: NREL 5.0 MW Baseline Wind Turbine (Onshore)

 Running ElastoDyn (v1.04.00a-bjj, 26-Jul-2016).

 Running AeroDyn (v15.03.00, 27-Jul-2016).

 Running AirfoilInfo (v1.01.00a-bjj, 5-Apr-2016).

 Running BEM (v1.02.00, 29-Jun-2016).
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 1
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 1
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 1
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 1
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 2
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 2
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 2
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 2
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 3
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 3
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 3
Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 3

 Running InflowWind (v3.03.00, 26-Jul-2016).
 Opening InflowWind input file:  .\5MW_Baseline/NRELOffshrBsline5MW_InflowWind_12mps.dat

   Reading a 31x31 grid (145 m wide, 17.5 m to 162.5 m above ground) with a characteristic wind
   speed of 12 m/s. This full-field file was generated by TurbSim (v1.06.00, 21-Sep-2012) on
   07-Jan-2014 at 12:50:45.

   Processed 1442 time steps of 20-Hz full-field data (72.05 seconds).

 Running ServoDyn (v1.06.00a-bjj, 26-Jul-2016).

 Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows,
 14-Oct-2015).
 Timestep: 0 of 60 seconds.

FAST_Solution0:CalcOutputs_And_SolveForInputs:SolveOption2:SrvD_CalcOutput:Running with torque
and pitch control of the NREL offshore 5MW baseline wind turbine from DISCON.dll as written by J.
Jonkman of NREL/NWTC for use in the IEA Annex XXIII OC3 studies.


 Total Real Time:       25.584 seconds
 Total CPU Time:        25.188 seconds
 Simulation CPU Time:   19.828 seconds
 Simulated Time:        60 seconds
 Time Ratio (Sim/CPU):  3.026

 FAST terminated normally.

**Comment by TheAfzalAmanullah** [comment]
I changed in **NRELOffshrBsline5MW_Onshore_ServoDyn.dat** file as you suggested and then edit in the file  **Run_OpenLoop.m** at  this place **for iTest = [18:18]**  and finally run the **Run_OpenLoop.m** script file.  I got the following output.

Run_OpenLoop
***********************************************
FAST_SFunc certification test for Test18
***********************************************

 **************************************************************************************************
 FAST (v8.16.00a-bjj, 27-Jul-2016)

 Copyright (C) 2016 National Renewable Energy Laboratory

 This program comes with ABSOLUTELY NO WARRANTY. See the "license.txt" file distributed with this
 software for details.
 **************************************************************************************************

  Running FAST (v8.16.00a-bjj, 27-Jul-2016), compiled as a DLL S-Function for Simulink as a 64-bit
  application using single precision
  linked with NWTC Subroutine Library (v2.09.00, 23-Jul-2016)

  Heading of the FAST input file:
    FAST Certification Test #18: NREL 5.0 MW Baseline Wind Turbine (Onshore)

  Running ElastoDyn (v1.04.00a-bjj, 26-Jul-2016).

  Running AeroDyn (v15.03.00, 27-Jul-2016).

  Running AirfoilInfo (v1.01.00a-bjj, 5-Apr-2016).

  Running BEM (v1.02.00, 29-Jun-2016).
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 1
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 1
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 1
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 1
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 2
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 2
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 2
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 2
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 1, Blade = 3
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 2, Blade = 3
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 3, Blade = 3
 Warning: Turning off Unsteady Aerodynamics because C_nalpha is 0.  BladeNode = 4, Blade = 3

  Running InflowWind (v3.03.00, 26-Jul-2016).
  Opening InflowWind input file:
  ..\..\CertTest\5MW_Baseline/NRELOffshrBsline5MW_InflowWind_12mps.dat

    Reading a 31x31 grid (145 m wide, 17.5 m to 162.5 m above ground) with a characteristic wind
    speed of 12 m/s. This full-field file was generated by TurbSim (v1.06.00, 21-Sep-2012) on
    07-Jan-2014 at 12:50:45.

    Processed 1442 time steps of 20-Hz full-field data (72.05 seconds).

  Running ServoDyn (v1.06.00a-bjj, 26-Jul-2016).

  Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for
  Windows/Matlab, 14-Oct-2015).

FAST_InitializeAll:SrvD_Init:ValidatePrimaryData:Yaw angle and rate are not commanded from Simulink model.
ValidatePrimaryData:Pitch angles are not commanded from Simulink model.
ValidatePrimaryData:Generator torque and power are not commanded from Simulink model.
ValidatePrimaryData:HSS brake is not commanded from Simulink model.
  Timestep: 0 of 60 seconds.

  FAST completed.

**Reply by borgestassio** [reply]
Since you could successfully run the Test18.fst, I suggest you to continue with the tutorial.
All controllers presented in this repository are simulated using a modification of Test18.fst.
If you want to know where the error is coming from, I suggest you to run each test individually and analyze the output log, as the error only seems to be related to some configuration in either a .fst or a .dat file somewhere.

**Reply by TheAfzalAmanullah** [reply]
I went up to Test26.fst, Run_OpenLoop.m file ran completely fine except for  (_Test19.fst - Test25.fst_). For (19-25) Test, it create same error.

Error using Run_OpenLoop (line 30)
Error reported by S-function 'FAST_SFunc' in 'OpenLoop/FAST Nonlinear Wind Turbine/S-Function':
FAST_InitializeAll:SrvD_Init:ValidatePrimaryData:Yaw angle and rate are not commanded from Simulink model.
ValidatePrimaryData:Pitch angles are not commanded from Simulink model.
ValidatePrimaryData:Generator torque and power are not commanded from Simulink model.
ValidatePrimaryData:HSS brake is not commanded from Simulink model.
SrvD_Init:BladedInterface_Init:The dynamic library ..\..\CertTest\5MW_Baseline/ServoData/DISCON_win32.dll could not be loaded. Check that the file exists in the specified location and that it is
compiled for 64-bit applications.

What should I do to remove this error?
Thank you

**Reply by borgestassio** [reply]
Alright, now I understand what is causing the problem.
I forgot to include the instructions to change the DLL file for those tests as well, my apologies.
So, in order to fix that.
Go to each "**...ServoDyn.da**t" file on the folder "**5MW_Baseline**" and change the line 66 to the 64bit DLL as well.

I will update the instructions on the readme

**Reply by TheAfzalAmanullah** [reply]
Now, finally this script  **Run_OpenLoop.m** working fine. Thanks a lot. After running this file I got 7 output file for all 26 Test. For example here I mentioned for Test26. How can I realize of the output file for all this case since _Test26.SFunc.outb_  is not understandable? Please help me to realize the output. 
1. Test26.SFunc.AD.sum, 
2. Test26.SFunc.BD1.sum
3. Test26.SFunc.BD2.sum
4. Test26.SFunc.BD3.sum
5. Test26.SFunc.ED.sum
6. Test26.SFunc.sum
7. Test26.SFunc.outb

**Reply by TheAfzalAmanullah** [reply]
I used the command  `ReadFASTbinary('Test18.outb')`  , but still getting error.

ReadFASTbinary('Test18.outb')
Error using ReadFASTbinary (line 113)
Could not open the FAST binary file: Test18.outb

**Reply by borgestassio** [reply]
The .sum files are the summaries for each component.
The outb is the output in binary.

**Reply by borgestassio** [reply]
if you want a human-readable output file, you can change the line 39 on the .fst file to '1' instead of '2'.
Note that the file will be significantly larger.

**Reply by borgestassio** [reply]
> I used the command `ReadFASTbinary('Test18.outb')` , but still getting error.
> 
> ReadFASTbinary('Test18.outb') Error using ReadFASTbinary (line 113) Could not open the FAST binary file: Test18.outb

Are you sure you're executing the code while in the folder where the file is?
if you can see the outb file and it's on your Matlab current workspace, I don't see a reason it wouldn't work.

**Comment by TheAfzalAmanullah** [comment]
I downloaded a file [5MW](https://github.com/borgestassio/Wind-Turbine-Control/blob/master/5MW/5MW.rar) and then run it. I got again an error. What is the reason behind this?
Error using testcontrole5mw (line 61)
Error reported by S-function 'FAST_SFunc' in 'TestControle/FAST Nonlinear Wind Turbine/S-Function':
FAST_Solution:CalcOutputs_And_SolveForInputs:SolveOption2:InflowWind_CalcOutput:CalcOutput:IfW_TSFFWind_CalcOutput [position=(-9.1263, -16.491, 44.504) in wind-file coordinates]: Error: FF wind array
was exhausted at 65.3 seconds (trying to access data at 72.102 seconds).

**Reply by borgestassio** [reply]
Yes, it was something I should have fixed already.
The wind files has only 60s of wind data, therefore if you run a simulation with TMax > 60 this error will occur.
You can change the TMax to 60 and it should run correctly.

**Reply by TheAfzalAmanullah** [reply]
Thanks a lot for your reply sir. This file is now working fine.
Thank you
